models/hourly_spend.sql

```diff
@@ -0,0 +1,429 @@
+-- depends_on: {{ ref('stg_metering_history') }}
+{{ config(materialized='table') }}
+
+with hour_spine as (
+    {% if execute %}
+{% set stg_metering_history_relation = load_relation(ref('stg_metering_history')) %}
+        {% if stg_metering_history_relation %}
+            {% set results = run_query("select min(convert_timezone('UTC', start_time)) from " ~ ref('stg_metering_history')) %}
+            {% set start_date = "'" ~ results.columns[0][0] ~ "'" %}
+            {% set results = run_query("select dateadd(hour, 1, max(convert_timezone('UTC', start_time))) from " ~ ref('stg_metering_history')) %}
+            {% set end_date = "'" ~ results.columns[0][0] ~ "'" %}
+        {% else %}
+            {% set start_date = "'2023-01-01 00:00:00'" %} {# this is just a dummy date for initial compilations before stg_metering_history exists #}
+            {% set end_date = "'2023-01-01 01:00:00'" %} {# this is just a dummy date for initial compilations before stg_metering_history exists #}
+        {% endif %}
+    {% endif %}
+{{ dbt_utils.date_spine(
+            datepart="hour",
+            start_date=start_date,
+            end_date=end_date
+        )
+    }}
+),
+
+hours as (
+    select
+        date_hour as hour,
+        hour::date as date,
+        day(last_day(date)) as days_in_month
+    from hour_spine
+),
+
+storage_terabytes_daily as (
+    select
+        date,
+        'Table and Time Travel' as storage_type,
+        database_name,
+        sum(average_database_bytes) / power(1024, 4) as storage_terabytes
+    from {{ ref('stg_database_storage_usage_history') }}
+    group by 1, 2, 3
+    union all
+    select
+        date,
+        'Failsafe' as storage_type,
+        database_name,
+        sum(average_failsafe_bytes) / power(1024, 4) as storage_terabytes
+    from {{ ref('stg_database_storage_usage_history') }}
+    group by 1, 2, 3
+    union all
+    select
+        date,
+        'Stage' as storage_type,
+        null as database_name,
+        sum(average_stage_bytes) / power(1024, 4) as storage_terabytes
+    from {{ ref('stg_stage_storage_usage_history') }}
+    group by 1, 2, 3
+),
```

Comment:
do these tables have records for the current day?

if they don't, storage values will be present for the latest day with 0 spend and users will need to handle that accordingly in their queries (we should call out as a caveat in the model docs as well).  previously daily spend filtered all input datasets to only go up to the last complete day so this wasn't an issue.

Reply:
They do.